### PR TITLE
Refactor network client constructor

### LINKED ISSUE #1047: Refactor network client construction
At the moment, the "network client" (i.e. the provider and block tracker for a specific network) is constructed in two separate places. The middleware and block tracker are constructed in the dedicated network client module (`createInfuraClient.js` or `createJsonRpcClient.js`), then the middleware is extended further in the network controller.

This arrangement makes it ambiguous where the network client starts and ends. Having a clear representation of the network client in our codebase is important for simultaneous network support.

We should refactor the network clients such that the entire network client is setup in a dedicated module, distinct from the network controller.

### PR DESCRIPTION
**Pull request by adonesky1**
Fixes: https://github.com/MetaMask/core/issues/1047

**Per Mark's description for the ticket this is addressing:**

_At the moment, the "network client" (i.e. the provider and block tracker for a specific network) is constructed in two separate places. The middleware and block tracker are constructed in the dedicated network client module (createInfuraClient.js or createJsonRpcClient.js), then the middleware is extended further in the network controller._

_This arrangement makes it ambiguous where the network client starts and ends. Having a clear representation of the network client in our codebase is important for simultaneous network support._

This PR consolidates `createInfuraClient` and `createJsonRpcClient` and the assembly  middleware that each client requires into a single module in order to improve legibility, maintainability and extensibility as we move towards supporting concurrent multi-chain request routing.

### CODE REVIEW

File: app/scripts/controllers/network/provider-api-tests/helpers.js [hunk 1]
```diff
@@ -3,8 +3,7 @@ import sinon from 'sinon';
 import { JsonRpcEngine } from 'json-rpc-engine';
 import { providerFromEngine } from 'eth-json-rpc-middleware';
 import EthQuery from 'eth-query';
-import createInfuraClient from '../createInfuraClient';
-import createJsonRpcClient from '../createJsonRpcClient';
+import { createNetworkClient } from '../create-network-client';
```
Comment (BelfordZ) [inline app/scripts/controllers/network/provider-api-tests/helpers.js:6]: *chefs kiss*

File: app/scripts/controllers/network/provider-api-tests/shared-tests.js [hunk 2]
```diff
@@ -117,29 +117,30 @@ export function testsForProviderType(providerType) {
         { name: 'eth_coinbase', numberOfParameters: 0 },
         { name: 'eth_createAccessList', numberOfParameters: 2 },
         { name: 'eth_feeHistory', numberOfParameters: 3 },
-        { name: 'eth_getFilterChanges', numberOfParameters: 1 },
```
Comment (mcmire) [inline app/scripts/controllers/network/provider-api-tests/shared-tests.js:120]: Why were these removed in this PR? Shuffling code around shouldn't affect the tests.

File: package.json [hunk 3]
```diff
@@ -275,13 +276,14 @@
...
     "bowser": "^2.11.0",
+    "btoa": "^1.2.1",
     "classnames": "^2.2.6",
     "copy-to-clipboard": "^3.0.8",
     "currency-formatter": "^1.4.2",
     "debounce-stream": "^2.0.0",
     "deep-freeze-strict": "1.1.1",
     "end-of-stream": "^1.4.4",
-    "eth-block-tracker": "^6.0.0",
+    "eth-block-tracker": "7.0.0",
```
Comment (mcmire) [inline package.json:286]: Can we use `^` here?

File: package.json [hunk 4]
```diff
@@ -229,8 +229,9 @@
     "@metamask/controller-utils": "^1.0.0",
     "@metamask/design-tokens": "^1.9.0",
     "@metamask/desktop": "^0.3.0",
-    "@metamask/eth-json-rpc-infura": "^7.0.0",
-    "@metamask/eth-json-rpc-middleware": "^10.0.0",
+    "@metamask/eth-json-rpc-infura": "8.0.0",
```
Comment (mcmire) [inline package.json:232]: Can we use `^` here?

File: app/scripts/controllers/network/create-network-client.ts [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    code: number;
+  };
+};
+
+function createNetworkAndChainIdMiddleware({
+  network,
+}: {
+  network: InfuraSupportedNetwork;
+}) {
+  if (!BUILT_IN_NETWORKS[network]) {
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts:51]: yeah that all makes sense to me!

File: app/scripts/controllers/network/create-network-client.ts [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+ * @returns
+ */
+export function createNetworkClient(
+  networkConfig: CustomNetworkConfiguration | InfuraNetworkConfiguration,
+): { provider: SafeEventEmitterProvider; blockTracker: PollingBlockTracker } {
+  /* eslint-disable @typescript-eslint/no-require-imports,@typescript-eslint/no-shadow */
+  const fetch = global.fetch || require('node-fetch');
+  const btoa = global.btoa || require('btoa');
+
+  const rpcApiMiddleware: JsonRpcMiddleware<unknown, unknown> =
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts:141]: Yeah... not sure why I added this

File: app/scripts/controllers/network/create-network-client.ts [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+
+  engine.push(networkMiddleware);
+
+  const provider = providerFromEngine(engine);
+
+  return { provider, blockTracker };
+}
+
+function createChainIdMiddleware(chainId: string) {
+  return (
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts:187]: 🤔 

like this?
```
const createChainIdMiddleware: JsonRpcMiddleware<unknown, unknown> = (
  chainId: string,
) => {
  return (
    req: any,
    res: any,
    next: () => Promise<void>,
    end: () => Promise<void>,
  ) => {
    if (req.method === 'eth_chainId') {
      res.result = chainId;
      return end();
    }
    return next();
  };
};
```

File: app/scripts/controllers/network/provider-api-tests/shared-tests.js (hunk 2 above)
Comment (adonesky1) [reply app/scripts/controllers/network/provider-api-tests/shared-tests.js:71]: These are removed before [we reverted that move subscribe and filter middleware commit](https://github.com/MetaMask/metamask-extension/pull/18129), since those middleware were no longer separable from the network client and we were going to have to put in bunch of effort to test them and we [were separating that effort out](https://github.com/MetaMask/metamask-extension/issues/16899). But now that that was reverted we can add them back!

File: package.json (hunk 4 above)
Comment (adonesky1) [reply package.json]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: package.json (hunk 3 above)
Comment (adonesky1) [reply package.json]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/provider-api-tests/shared-tests.js (hunk 2 above)
Comment (adonesky1) [reply app/scripts/controllers/network/provider-api-tests/shared-tests.js]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/create-network-client.ts [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+  networkConfig: CustomNetworkConfiguration | InfuraNetworkConfiguration,
+): { provider: SafeEventEmitterProvider; blockTracker: PollingBlockTracker } {
+  /* eslint-disable @typescript-eslint/no-require-imports,@typescript-eslint/no-shadow */
+  const fetch = global.fetch || require('node-fetch');
+  const btoa = global.btoa || require('btoa');
+
+  const rpcApiMiddleware: JsonRpcMiddleware<unknown, unknown> =
+    networkConfig.type === NetworkClientType.INFURA
+      ? createInfuraMiddleware({
+          // InfuraSupportedNetwork is a subset of InfuraJsonRpcSupportedNetwork
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: yea! `sepolia` was missing in v7 of `@metamask/eth-json-rpc-infura` but is now included so this works without casting.

done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/create-network-client.ts [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+  id: Y['id'];
+  jsonrpc: Y['jsonrpc'];
+  result: V | undefined;
+  error?: {
+    message: unknown;
+    code: number;
+  };
+};
+
+function createNetworkAndChainIdMiddleware({
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/create-network-client.ts [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+    createBlockRefMiddleware({ blockTracker, provider: rpcProvider }),
+    createRetryOnEmptyMiddleware({ blockTracker, provider: rpcProvider }),
+    createBlockTrackerInspectorMiddleware({ blockTracker }),
+    rpcApiMiddleware,
+  ]);
+}
+
+export enum NetworkClientType {
+  CUSTOM = 'custom',
+  INFURA = 'infura',
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/create-network-client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/4e3911a00695766a0edef23a4958edc66779bb24

File: app/scripts/controllers/network/create-network-client.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (mcmire) [reply app/scripts/controllers/network/create-network-client.ts:147]: Almost! Let me leave a suggestion.

File: app/scripts/controllers/network/create-network-client.ts [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+    next: () => Promise<void>,
+    end: () => Promise<void>,
+  ) => {
+    if (req.method === 'eth_chainId') {
+      res.result = chainId;
+      return end();
+    }
+    return next();
+  };
+};
```
Comment (mcmire) [inline app/scripts/controllers/network/create-network-client.ts:159]: The `createChainIdMiddleware` function doesn't represent the middleware, it returns the middleware. So you should be able to say:

``` suggestion
const createChainIdMiddleware = (chainId: string) => {
  const chainIdMiddleware: JsonRpcMiddleware<unknown, unknown> (req, res, next, end) => {
    if (req.method === 'eth_chainId') {
      res.result = chainId;
      return end();
    }
    return next();
  };
  return chainIdMiddleware;
};
```

I guess now that I've looked at your example you can also move the type this way:

``` suggestion
const createChainIdMiddleware = (chainId: string): JsonRpcMiddleware<unknown, unknown> => {
  return (req, res, next, end) => {
    if (req.method === 'eth_chainId') {
      res.result = chainId;
      return end();
    }
    return next();
  };
};
```

Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/45e028e815c9fbb7180d10e902dd130255afe740

File: app/scripts/controllers/network/create-network-client.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: done here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/45e028e815c9fbb7180d10e902dd130255afe740

**Review by mcmire — DISMISSED** [review]
Looks good to me! Nice work 😎

File: package.json [hunk 12]
```diff
@@ -279,13 +280,14 @@
     "bignumber.js": "^4.1.0",
     "bn.js": "^4.11.7",
     "bowser": "^2.11.0",
+    "btoa": "^1.2.1",
```
Comment (Gudahtt) [inline package.json:283]: Nit: If we set this during the unit test setup script, we can move this to `devDependencies` and avoid having this in our bundle

File: app/scripts/controllers/network/create-network-client.ts [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+/**
+ * Create a JSON RPC network client for a specific network.
+ *
+ * @param networkConfig - The network configuration.
+ * @returns
+ */
+export function createNetworkClient(
+  networkConfig: CustomNetworkConfiguration | InfuraNetworkConfiguration,
+): { provider: SafeEventEmitterProvider; blockTracker: PollingBlockTracker } {
+  /* eslint-disable @typescript-eslint/no-require-imports,@typescript-eslint/no-shadow */
```
[the remark this answers is not in the document]
Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts:57]: seems like tests are working without setting fallback `btoa` in setup? https://github.com/MetaMask/metamask-extension/pull/17652/commits/5bb475037056a114cec96db35fbc52e503a8fcc4

File: package.json (hunk 12 above)
Comment (adonesky1) [reply package.json]: removed from deps here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/5bb475037056a114cec96db35fbc52e503a8fcc4

File: app/scripts/controllers/network/create-network-client.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply app/scripts/controllers/network/create-network-client.ts:57]: Oh right, I forgot, it was added to Node.js in v16: https://nodejs.org/docs/latest-v16.x/api/globals.html#btoadata
Cool, even better!

File: app/scripts/controllers/network/create-network-client.ts [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+  const fetch = global.fetch || require('node-fetch');
```
Comment (Gudahtt) [inline app/scripts/controllers/network/create-network-client.ts:58]: We're in a similar situation here; we can drop the fallback because `fetch` is supported by all of our supported browsers. We have it setup in unit tests already as well.

Comment (adonesky1) [reply app/scripts/controllers/network/create-network-client.ts]: removed here: https://github.com/MetaMask/metamask-extension/pull/17652/commits/2c67287e814d42f8c80161560ed966b4008e924f

**Review by Gudahtt — APPROVED** [review]
LGTM!

**Review by mcmire — APPROVED** [review]
